src/storage-operators/src/s3_oneshot_sink/parquet.rs

```diff
@@ -7,28 +7,367 @@
+/// │ │ColBuildr│ │ColBuildr│ ├──┼─┼──►│  buffer  │   │       │
+/// │ │         │ │         │ │  │ │   └─────┬────┘   │       │
+/// │ └─────────┘ └─────────┘ │  │ │         │        │       │
+/// │                         │  │ │   ┌─────▼────┐   │       │
+/// └─────────────────────────┘  │ │   │ row group│   │       │
+///                              │ │   └─┬────────┘   │       │
+///                              │ │     │            │       │
+///                              │ └─────┼────────────┘       │
+///                              │  ┌────┼──────────────────┐ │
+///                              │  │    │       S3MultiPart│ │
+///                              │  │ ┌──▼───────┐ Uploader │ │
+///                              │  │ │  buffer  │          │ │
+///      ┌─────────┐             │  │ └───┬─────┬┘          │ │
+///      │ S3 API  │◄────────────┼──┤     │     │           │ │
+///      └─────────┘             │  │ ┌───▼──┐ ┌▼─────┐     │ │
+///                              │  │ │ part │ │ part │     │ │
+///                              │  │ └──────┘ └──────┘     │ │
+///                              │  │                       │ │
+///                              │  └───────────────────────┘ │
+///                              │                            │
+///                              └────────────────────────────┘
+///
+/// ## File Size & Buffer Sizes
+///
+/// We expose a 'MAX FILE SIZE' parameter to the user, but this is difficult to enforce exactly
+/// since we don't know the exact size of the data we're writing before a parquet row-group
+/// is flushed. This is because the encoded size of the data is different than the in-memory
+/// representation and because the data pages within each column in a row-group are compressed.
+/// We also don't know the exact size of the parquet metadata that will be written to the file.
+///
+/// Therefore we don't use the S3MultiPartUploader's hard file size limit since it's difficult
+/// to handle those errors after we've already flushed data to the ArrowWriter. Instead we
+/// implement a crude check ourselves.
+///
+/// This check aims to hit the max-size limit but may exceed it by some amount. To ensure
+/// that amount is small, we set the max row-group size to a configurable ratio (e.g. 20%)
+/// of the max_file_size. This determines how often we'll flush a row-group, but is only an
+/// approximation since the actual size of the row-group is not known until it's written.
+/// After each row-group is flushed, the size of the file is checked and if it's exceeded
+/// max-file-size a new file is started.
+///
+/// We also set the max ArrowBuilder buffer size to a ratio (e.g. 150%) of the row-group size
+/// to avoid the ArrowWriter buffering too much data itself before flushing a row-group. We're
+/// aiming for the encoded & compressed size of the ArrowBuilder data to be roughly equal to
+/// the row-group size, but this is only an approximation.
```

Comment:
what happens if we go over this 150% limit?

Reply:
After the ArrowBuilder gets converted to a RecordBatch and this `RecordBatch` is encoded to a size much larger than the max row-group size, it'll just result in us creating an extra large row-group in the output file. If this happens as the last row-group to be added to the file (since the file goes over the file-size limit), then this could result in us exceeding the max file size by a non-small amount.

I do expect we'll have to tune this ratio to make it work well. I'm going to add an issue to our epic to introduce some test cases that help us tune this parameter by validating how much we end up exceeding the file size limits on large data dumps.